File: frontend/src/features/production/api/procedureTagPrices.ts
import { service } from '@/api/request'

export type ProcedureTagPriceTag = {
  id: number
  tag_name: string
  unit_price: number | string | null
}

export type ProcedureTagPriceProcedure = {
  procedure_id: number
  procedure_name: string
  tags_locked: boolean
  available_tags: ProcedureTagPriceTag[]
  configured_tags: ProcedureTagPriceTag[]
}

export type ProcedureTagPricePart = {
  product_id: number
  product_version: number
  product_name: string
  factory_code: string
  product_bom_id: number | null
  origin_flow_node_id: string
  part_name: string
  part_no: string
  procedures: ProcedureTagPriceProcedure[]
}

export async function queryProcedureTagPrices(
  departmentCode: string,
  page: number,
  pageSize: number,
  keyword: string,
) {
  const response = await service.get<{
    data: ProcedureTagPricePart[]
    total: number
  }>(`/departments/${departmentCode}/procedure-tag-prices`, {
    params: {
      page,
      page_size: pageSize,
      keyword: keyword.trim() || undefined,
    },
  })
  return response.data
}

export async function saveProcedureTagPrices(
  departmentCode: string,
  productId: number,
  productVersion: number,
  originFlowNodeId: string,
  procedureId: number,
  tags: Array<{ tag_name: string; unit_price: number | null }>,
) {
  await service.put(
    `/departments/${departmentCode}/procedure-tag-prices/${productId}/${productVersion}/${encodeURIComponent(originFlowNodeId)}/${procedureId}`,
    { tags },
  )
}

File: backend/src/schemas/procedure_tag_prices.py
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field


class ProcedureTagPriceModel(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)


class ProcedureTagPriceTagResponse(ProcedureTagPriceModel):
    id: int
    tag_name: str
    unit_price: Decimal | None


class ProcedureTagPriceProcedureResponse(ProcedureTagPriceModel):
    procedure_id: int
    procedure_name: str
    tags_locked: bool
    available_tags: list[ProcedureTagPriceTagResponse]
    configured_tags: list[ProcedureTagPriceTagResponse]


class ProcedureTagPricePartResponse(ProcedureTagPriceModel):
    product_id: int
    product_version: int
    product_name: str
    factory_code: str
    product_bom_id: int | None
    origin_flow_node_id: str
    part_name: str
    part_no: str
    procedures: list[ProcedureTagPriceProcedureResponse]


class ProcedureTagPriceListEnvelope(ProcedureTagPriceModel):
    data: list[ProcedureTagPricePartResponse]
    total: int


class ProcedureTagPriceInput(ProcedureTagPriceModel):
    tag_name: str = Field(min_length=1, max_length=200)
    unit_price: Decimal | None = Field(default=None, ge=0, max_digits=12, decimal_places=2)


class ProcedureTagPriceUpdate(ProcedureTagPriceModel):
    tags: list[ProcedureTagPriceInput] = Field(default_factory=list, max_length=100)
